src/secrets/actions/thunks.ts

```diff
@@ -0,0 +1,117 @@
+// Libraries
+import {normalize} from 'normalizr'
+
+// API
+import {
+  getOrgsSecrets as apiGetSecrets,
+  patchOrgsSecrets as apiUpdateSecret,
+  postOrgsSecretsDelete as apiDeleteSecret,
+} from 'src/client'
+
+// Schemas
+import {secretsSchema, arrayOfSecrets} from 'src/schemas/secrets'
+
+// Types
+import {Dispatch} from 'react'
+import {
+  RemoteDataState,
+  GetState,
+  ResourceType,
+  Secret,
+  SecretEntities,
+} from 'src/types'
+
+// Actions
+import {notify} from 'src/shared/actions/notifications'
+import {
+  getSecretsFailed,
+  upsertSecretFailed,
+  deleteSecretsFailed,
+} from 'src/shared/copy/notifications'
+import {
+  setSecrets,
+  setSecret,
+  removeSecret,
+  Action,
+} from 'src/secrets/actions/creators'
+
+// Utils
+import {getOrg} from 'src/organizations/selectors'
+import {getStatus} from 'src/resources/selectors'
+
+const makeEntitiesForSecrets = (response: Array<string>) => {
+  const processedData = []
+  response.forEach(element => processedData.push({id: element, key: element}))
+  return processedData
+}
+
+export const getSecrets = () => async (
+  dispatch: Dispatch<Action>,
+  getState: GetState
+) => {
+  const state = getState()
+  if (getStatus(state, ResourceType.Secrets) === RemoteDataState.NotStarted) {
+    dispatch(setSecrets(RemoteDataState.Loading))
+  }
+
+  const org = getOrg(state)
+
+  const resp = await apiGetSecrets({orgID: org.id})
+
+  if (resp.status === 200) {
+    const secrets = makeEntitiesForSecrets(resp.data.secrets)
+    const test = normalize<Secret, SecretEntities, string[]>(
+      secrets,
+      arrayOfSecrets
+    )
+    dispatch(setSecrets(RemoteDataState.Done, test))
+  } else {
+    console.error(resp.data.message)
+    dispatch(setSecrets(RemoteDataState.Error))
+    dispatch(notify(getSecretsFailed()))
+  }
+}
+
+export const upsertSecret = (_secretKey: string, secretValue: string) => async (
+  dispatch: Dispatch<Action>,
+  getState: GetState
+): Promise<void> => {
+  const org = getOrg(getState())
+  const resp = await apiUpdateSecret({
+    orgID: org.id,
+    data: {
+      _secretKey: secretValue,
+    },
+  })
+
+  if (resp.status === 204) {
+    const secret = normalize<Secret, SecretEntities, string>(
+      resp.data.secret,
+      secretsSchema
+    )
+    dispatch(setSecret(resp.data.secret.id, RemoteDataState.Done, secret))
+  } else {
+    console.error(resp.data.message)
+    dispatch(notify(upsertSecretFailed()))
+  }
+}
+
+export const deleteSecret = (id: string) => async (
+  dispatch: Dispatch<Action>,
+  getState: GetState
+) => {
+  const org = getOrg(getState())
+  const resp = await apiDeleteSecret({
+    orgID: org.id,
+    data: {
+      secrets: [id],
```

Comment:
why is the API expecting an array of secrets to be deleted? Can we update the API definition here since we're only ever going to be deleting one at a time right?

Reply:
The generatedRoutes.ts expected an array in `PostOrgsSecretsDeleteParams` and I wasn't sure how to go about changing it so I just rolled with what was there.  When I finish implementing deletion next week it'll always be an array with a single element as the UI has a single delete button on a given secret.  If there's a way to change that around easily from our end I don't mind doing it I just wanted to avoid involving multiple people to change the definition from a single element array to a single element.

Reply:
For sure, I was just curious if this endpoint is being used by the CLI and is built out that way or if the swagger definition and API can be updated to consolidate this behavior since it seems a bit weird to me. We've run into this in several places so I know it's not an issue for you to resolve, just curious if this was designed with the intent that an array of secret ids can be passed in or if it's only ever expecting 1 and we can update the API to address that discrepancy

Reply:
The docs [here](https://docs.influxdata.com/influxdb/v2.0/security/secrets/manage-secrets/delete/) say (emphasis mine):
> Use the POST request method and the orgs/{orgID}/secrets/delete API endpoint to delete one **or more** secrets.

The CLI only seems to allow one delete at a time as far as I can tell (I've tried and the docs don't say "or more" there).  I agree it's a bit strange but it seems intentional so I was just going to roll with it and pass in single element arrays.

Reply:
@glinton it looks like the openAPI description here is that it's to delete a secret, but the $ref is to Secrets. Should we change the schema definition to `Secret` or update the description?

https://github.com/influxdata/openapi/blob/50ceec33e0c114d39a9399efd9a0929aa54c77b4/src/common/paths/orgs_orgID_secrets_delete.yml#L14-L20

Reply:
so glad this came up. i'm not sure who added this route to the original api, but i'd love for it to change if ya'll are working on it. we can mark the one that posts to a `/delete` route as deprecated and add a `DELETE /secrets/{id}` route. The original looked to be a workaround anyway ([`"/orgs/{orgID}/secrets/delete": # had to make this because swagger wouldn't let me have a request body with a DELETE`](https://raw.githubusercontent.com/influxdata/influxdb/b66079f08470575dc1b22c80be2a6be39f6e7b4c/http/swagger.yml)).  Not knowing really what that api is used for and assuming that the keys currently passed in are actually the ids of the key, i think it would be fine to do a `delete` request to the secret url.

Regarding the question about updating the schema, I'm seeing `SecretKeys` (which as mentioned above, is ambiguous)
![image](https://user-images.githubusercontent.com/2653109/122476508-91e06600-cf83-11eb-89bf-9c70d4338909.png)

I don't remember ever checking, but i've believed that the api accepts (and requires) an array of strings.

Reply:
@glinton Implementing the delete is on my to do list for next week so this is definitely being worked on.  The spec as written has us only ever deleting one secret at a time so a `DELETE /secrets/{id}` route makes a lot of sense to me.

Reply:
if you need any assistance building support for it on the api side, feel free to reach out